File: src/android/Xfyun.java
package org.cordova.plugin.xfyun;

import android.content.Context;
import android.util.Log;
import android.os.Bundle;
import android.Manifest;
import android.content.pm.PackageManager;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.iflytek.cloud.ErrorCode;
import com.iflytek.cloud.GrammarListener;
import com.iflytek.cloud.InitListener;
import com.iflytek.cloud.LexiconListener;
import com.iflytek.cloud.RecognizerListener;
import com.iflytek.cloud.RecognizerResult;
import com.iflytek.cloud.SpeechConstant;
import com.iflytek.cloud.SpeechError;
import com.iflytek.cloud.SpeechRecognizer;
import com.iflytek.cloud.util.ContactManager;
import com.iflytek.cloud.util.ContactManager.ContactListener;
import com.iflytek.cloud.util.ResourceUtil;
import com.iflytek.cloud.util.ResourceUtil.RESOURCE_TYPE;
import com.iflytek.cloud.SpeechUtility;

/**
 * cordova科大讯飞插件
 * create by wilhan.tian
 */
public class Xfyun extends CordovaPlugin {
    private SpeechRecognizer mAsr;
    private String TAG = "cordova-plguin-xfyun";
    private CallbackContext mBuildGrammarCallbackContext;

    private CallbackContext mNoPerGrammarListeningCallbackContext;
    private JSONArray mNoPerGrammarArgs;

    private String [] permissions = { 
        Manifest.permission.RECORD_AUDIO, 
        // Manifest.permission.WRITE_EXTERNAL_STORAGE,
        // Manifest.permission.READ_EXTERNAL_STORAGE
    };

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("init")) {
            this.init(args, callbackContext);
            return true;
        }
        if (action.equals("buildGrammar")) {
            this.buildGrammar(args, callbackContext);
            return true;
        }
        if (action.equals("startListeningGrammar")) {
            this.startListeningGrammar(args, callbackContext);
            return true;
        }
        if (action.equals("stopListeningGrammar")) {
            this.stopListeningGrammar(args, callbackContext);
            return true;
        }
        if (action.equals("cancelGrammar")) {
            this.cancelGrammar(args, callbackContext);
            return true;
        }
        
        return false;
    }

    // 初始化
    private void init(JSONArray args, CallbackContext callbackContext) throws JSONException {
        String appid = args.getString(0);
        String engine_mode = args.getString(1);

        if(engine_mode == null || engine_mode.isEmpty()){
            engine_mode = "auto";
        }

        Context context = cordova.getActivity().getApplicationContext();
        SpeechUtility.createUtility(context, "appid=" + appid + ",engine_mode=" + engine_mode);
        callbackContext.success();
    }

    // 构建命令
    private void buildGrammar(JSONArray args, CallbackContext callbackContext) throws JSONException {
        String cloudGrammar = args.getString(0);

        // cloudGrammar = "#ABNF 1.0 UTF-8;\n"+
        //                 "language zh-CN;\n"+
        //                 "mode voice;\n"+
        //                 "root $main;\n"+
        //                 "$main = $place1 到 $place2;\n"+
        //                 "$place1 = 上海|合肥;\n"+
        //                 "$place2 = 北京|武汉|南京|天津|东京;";

        Context context = cordova.getActivity().getApplicationContext();
        if(mAsr == null){
            mAsr = SpeechRecognizer.createRecognizer(context, null);
        }

        mAsr.setParameter(SpeechConstant.TEXT_ENCODING, "utf-8");
        int ret = mAsr.buildGrammar("abnf", cloudGrammar, new GrammarListener() {
            @Override
            public void onBuildFinish(String grammarId, SpeechError error) {
                if(error == null){
                    if(grammarId != null){ //构建语法成功，请保存grammarId用于识别
                        Log.d(TAG,"grammarId = " + grammarId);
                        mBuildGrammarCallbackContext.success(grammarId);
                    }
                    else{
                        Log.d(TAG,"未知错误");
                        mBuildGrammarCallbackContext.error("未知错误");
                    }
                }
                else{
                    Log.d(TAG,"语法构建失败,错误码:" + error.getErrorCode());
                    mBuildGrammarCallbackContext.error(error.getErrorCode());
                }
            }
        });

        if (ret != ErrorCode.SUCCESS){
            Log.d(TAG,"语法构建准备失败,错误码:" + ret);
            callbackContext.error(ret);
        }else{
            Log.d(TAG,"语法构建正常进行");
            mBuildGrammarCallbackContext = callbackContext;
        }
    }

    /// 监听命令
    private void startListeningGrammar(JSONArray args, CallbackContext callbackContext) throws JSONException {
        if(mAsr == null){
            callbackContext.error(-1);//"请先成功buildGrammar(构建命令)"
            return;
        }

        // 获取相关权限
        if(!hasPermisssion()){
            this.cordova.requestPermissions(this, 0, permissions);
            mNoPerGrammarArgs = args;
            mNoPerGrammarListeningCallbackContext = callbackContext;

            PluginResult result = new PluginResult(PluginResult.Status.NO_RESULT);
            result.setKeepCallback(true);
            callbackContext.sendPluginResult(result);
            return;
        }

        String grammarId = args.getString(0);
        Log.d(TAG,"-------- grammarId="+grammarId);

        mAsr.setParameter(SpeechConstant.ENGINE_TYPE, "cloud"); 
        mAsr.setParameter(SpeechConstant.CLOUD_GRAMMAR, grammarId);

        int ret = mAsr.startListening(new CustomRecognizerListener(callbackContext));
        if (ret != ErrorCode.SUCCESS) {
            Log.d(TAG,"启动监听器失败, 错误码: " + ret);
            callbackContext.error(ret);
        }
        else{
            PluginResult result = new PluginResult(PluginResult.Status.NO_RESULT);
            result.setKeepCallback(true);
            callbackContext.sendPluginResult(result);
        }
    }

    // 停止命令监听
    private void stopListeningGrammar(JSONArray args, CallbackContext callbackContext) throws JSONException {
        if(mAsr == null){
            Log.d(TAG,"mAsr为空，停止失败");
            callbackContext.error("当前无命令监听, 无法停止");
            return;
        }
        else{
            mAsr.stopListening();
            Log.d(TAG,"停止成功");
            callbackContext.success();
        }
    }

    // 取消命令
    private void cancelGrammar(JSONArray args, CallbackContext callbackContext) throws JSONException {
        if(mAsr == null){
            callbackContext.error("当前无命令监听, 无法取消");
            return;
        }
        else{
            mAsr.cancel();
            mAsr = null;
            callbackContext.success();
        }
    }

    @Override
    public boolean hasPermisssion() {
        for(String p : permissions)
        {
            if(!this.cordova.hasPermission(p))
                return false;
        }
        return true;
    }

    @Override
    public void requestPermissions(int requestCode){
        this.cordova.requestPermissions(this, requestCode, permissions);
    }

    @Override
    public void onRequestPermissionResult(int requestCode, String[] permissions, int[] grantResults) throws JSONException
    {
        if(mNoPerGrammarListeningCallbackContext != null) {
            for (int r : grantResults) {
                if (r == PackageManager.PERMISSION_DENIED) {
                    Log.e(TAG, "相关权限被拒绝!");
                    mNoPerGrammarListeningCallbackContext.error(-2);
                    return;
                }
            }
            this.startListeningGrammar(mNoPerGrammarArgs, mNoPerGrammarListeningCallbackContext);
        }
    }

    // 自定义监听器
    class CustomRecognizerListener implements RecognizerListener{
        private CallbackContext callbackContext;

        CustomRecognizerListener(CallbackContext callbackContext){
            this.callbackContext = callbackContext;
        }

        // 音量变化
        public void onVolumeChanged(int volume, byte[] data) {
            Log.d(TAG,"onVolumeChanged");
            try{
                JSONObject main = new JSONObject();
                main.put("action", "onVolumeChanged");

                JSONObject json = new JSONObject();
                json.put("volume", volume);
                json.put("data", data);
                main.put("data", json);

                PluginResult pResult = new PluginResult(PluginResult.Status.OK, main);
                pResult.setKeepCallback(true);
                callbackContext.sendPluginResult(pResult);
            }catch(JSONException e){}
        }

        // 返回结果
        public void onResult(final RecognizerResult result, boolean isLast) {
            Log.d(TAG,"onResult");
            try{
                JSONObject main = new JSONObject();
                main.put("action", "onResult");
                
                JSONObject json = new JSONObject();
                json.put("result", result == null ? "" : result.getResultString());
                json.put("isLast", isLast);
                main.put("data", json);

                PluginResult pResult = new PluginResult(PluginResult.Status.OK, main);
                pResult.setKeepCallback(true);
                callbackContext.sendPluginResult(pResult);
            }catch(JSONException e){}
        }

        // 开始说话
        public void onBeginOfSpeech() {
            Log.d(TAG,"onBeginOfSpeech");
            try{
                JSONObject main = new JSONObject();
                main.put("action", "onBeginOfSpeech");
                
                PluginResult pResult = new PluginResult(PluginResult.Status.OK, main);
                pResult.setKeepCallback(true);
                callbackContext.sendPluginResult(pResult);
            }catch(JSONException e){}
        }

        // 结束说话
        public void onEndOfSpeech() {
            Log.d(TAG,"onEndOfSpeech");
            try{
                JSONObject main = new JSONObject();
                main.put("action", "onEndOfSpeech");
                
                PluginResult pResult = new PluginResult(PluginResult.Status.OK, main);
                pResult.setKeepCallback(true);
                callbackContext.sendPluginResult(pResult);
            }catch(JSONException e){}
        }

        // 错误回调
        public void onError(SpeechError error) {
            Log.d(TAG,"发生错误，错误码:"+error.getErrorCode());
            callbackContext.error(error.getErrorCode());
        }

        // 件回调
        public void onEvent(int eventType, int arg1, int arg2, Bundle obj) {}
    }
}

File: src/ios/Xfyun.m
/********* Xfyun.m Cordova Plugin Implementation *******/
/********* create by wilhan.tian *******/

#import <Cordova/CDV.h>
#import <iflyMSC/iflyMSC.h>

@interface Xfyun : CDVPlugin<IFlySpeechRecognizerDelegate>{
    NSString* mGrammarListeningCommandId;
}

- (void)init:(CDVInvokedUrlCommand*)command;
- (void)buildGrammar:(CDVInvokedUrlCommand*)command;
- (void)startListeningGrammar:(CDVInvokedUrlCommand*)command;
- (void)stopListeningGrammar:(CDVInvokedUrlCommand*)command;
- (void)cancelGrammar:(CDVInvokedUrlCommand*)command;
@end

///==========================================================================================

@implementation Xfyun

// 初始化
- (void)init:(CDVInvokedUrlCommand*)command
{
    NSString* appid = [command.arguments objectAtIndex:0];
    NSString* engine_mode = [command.arguments objectAtIndex:1];
    
    NSString *initString = [[NSString alloc] initWithFormat:@"appid=%@,engine_mode=%@", appid, engine_mode];
    [IFlySpeechUtility createUtility:initString];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

// 构建命令
- (void)buildGrammar:(CDVInvokedUrlCommand*)command
{
    NSString* cloudGrammar = [command.arguments objectAtIndex:0];
    NSLog(cloudGrammar);
    
    IFlySpeechRecognizer* recognizer = [IFlySpeechRecognizer sharedInstance];
    recognizer.delegate = self;
    
    //开启候选结果
    [recognizer setParameter:@"1" forKey:@"asr_wbest"];
    //设置引擎类型，cloud 或者 local
    [recognizer setParameter:@"cloud" forKey:[IFlySpeechConstant ENGINE_TYPE]];
    //设置字符编码为 utf-8
    [recognizer setParameter:@"utf-8" forKey:[IFlySpeechConstant TEXT_ENCODING]];
    //语法类型，本地是 bnf，在线识别是 abnf
    [recognizer setParameter:@"abnf" forKey:[IFlyResourceUtil GRAMMARTYPE]];
    //设置服务类型为 asr 识别
    [recognizer setParameter:@"asr" forKey:[IFlySpeechConstant IFLY_DOMAIN]];
    //调用构建语法接口
    [recognizer buildGrammarCompletionHandler:^(NSString *grammarId, IFlySpeechError *error) {
        if(![error errorCode])
        {
            NSLog(@"构建语法成功 grammarId=%@", grammarId);
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:grammarId];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }
        else
        {
            NSLog(@"构建语法失败 errorCode=%d", [error errorCode]);
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsInt:[error errorCode]];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }
    } grammarType:@"abnf" grammarContent:cloudGrammar];
    
    // 保持回调
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_NO_RESULT];
    [pluginResult setKeepCallbackAsBool:TRUE];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

// 开始监听命令
- (void)startListeningGrammar:(CDVInvokedUrlCommand*)command
{
    NSString* grammerID = [command.arguments objectAtIndex:0];
    
    // 设置命令ID
    IFlySpeechRecognizer* recognizer = [IFlySpeechRecognizer sharedInstance];
    [recognizer setParameter:grammerID forKey:[IFlySpeechConstant CLOUD_GRAMMAR]];
    
    // 启动识别
    [recognizer startListening];
    
    // 保持回调
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_NO_RESULT];
    [pluginResult setKeepCallbackAsBool:TRUE];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    
    // 保存回调
    mGrammarListeningCommandId = command.callbackId;
    
    NSLog(@"开始监听");
}

// 停止监听
- (void)stopListeningGrammar:(CDVInvokedUrlCommand*)command
{
    IFlySpeechRecognizer* recognizer = [IFlySpeechRecognizer sharedInstance];
    
    if(recognizer.isListening)
    {
        [recognizer stopListening];
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
        
        NSLog(@"停止监听成功");
    }
    else
    {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"当前无命令监听, 无法停止"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        
        NSLog(@"停止监听失败，因为当前无监听");
    }
}

// 取消监听
- (void)cancelGrammar:(CDVInvokedUrlCommand*)command
{
    IFlySpeechRecognizer* recognizer = [IFlySpeechRecognizer sharedInstance];
    
    if(recognizer.isListening)
    {
        [recognizer cancel];
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
        
        NSLog(@"取消监听成功");
    }
    else
    {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"当前无命令监听, 无法取消"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        
        NSLog(@"取消监听失败，因为当前无监听");
    }
}

//--------------------------------------------------------------
//结果返回代理
- (void) onResults:(NSArray *) results isLast:(BOOL) isLast
{
    NSMutableDictionary* json = [[NSMutableDictionary alloc] init];
    [json setValue:@"onResult" forKey:@"action"];
    
    NSMutableDictionary* data = [[NSMutableDictionary alloc] init];
    [data setValue:results forKey:@"result"];
    [data setValue:isLast ? @(YES) : @(NO) forKey:@"isLast"];
    [json setValue:data forKey:@"data"];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:json];
    [pluginResult setKeepCallbackAsBool:TRUE];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:mGrammarListeningCommandId];
}
//音量回调
- (void) onVolumeChanged: (int)volume
{
    NSMutableDictionary* json = [[NSMutableDictionary alloc] init];
    [json setValue:@"onVolumeChanged" forKey:@"action"];
    
    NSMutableDictionary* data = [[NSMutableDictionary alloc] init];
    [data setValue:[NSNumber numberWithInt:volume] forKey:@"volume"];
    [json setValue:data forKey:@"data"];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:json];
    [pluginResult setKeepCallbackAsBool:TRUE];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:mGrammarListeningCommandId];
}
//录音开始回调
- (void) onBeginOfSpeech
{
    NSMutableDictionary* json = [[NSMutableDictionary alloc] init];
    [json setValue:@"onBeginOfSpeech" forKey:@"action"];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:json];
    [pluginResult setKeepCallbackAsBool:TRUE];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:mGrammarListeningCommandId];
}
//录音结束回调
- (void) onEndOfSpeech
{
    NSMutableDictionary* json = [[NSMutableDictionary alloc] init];
    [json setValue:@"onEndOfSpeech" forKey:@"action"];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:json];
    [pluginResult setKeepCallbackAsBool:TRUE];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:mGrammarListeningCommandId];
}
//会话错误结束回调
- (void) onError:(IFlySpeechError*) error
{
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsInt:[error errorCode]];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:mGrammarListeningCommandId];
}
//会话取消回调
- (void) onCancel{}
    
@end
